Warn on copying app if feature flags differ

orangejenny opened the pull request:
Just an idea after mulling over http://manage.dimagi.com/default.asp?240571 and some summit conversations, open to feedback.

@dannyroberts / @esoergel / @dimagi/product 

![screen shot 2016-11-02 at 12 10 59 am](https://cloud.githubusercontent.com/assets/1486591/19916528/2999db7e-a091-11e6-9bde-f63ca7f668de.png)

orangejenny committed:
warn on copying app if feature flags differ

dimagimon commented on corehq/apps/domain/views.py:
```diff
@@ -2525,6 +2528,20 @@ def page_context(self):
+    other_domain = params.get('domain')
+    diff = []
+    if Domain.get_by_name(other_domain):
+        diff = [(t.slug, t.label) for t in all_toggles() if t.enabled(request.domain) and not t.enabled(other_domain)]
```
E501 line too long (118 > 115 characters)

orangejenny committed:
[ci skip] lint

orangejenny committed:
work if support toggle is turned off

czue commented:
this is awesome. 

what do you think about providing the options "Copy and enable all" and "Copy without enabling"? I imagine that in most cases the person wants to enable everything, though providing that does make it easier for people to turn on lots of flags which may be against our long-term philosophy of trying to reduce flag usage...

amsagoff commented:
If this isn't just superuser facing, we may want to change the text
("feature flag" isn't a term external users are exposed to)

On Wed, Nov 2, 2016 at 9:23 AM, Cory Zue <notifications@github.com> wrote:

> this is awesome.
>
> what do you think about providing the options "Copy and enable all" and
> "Copy without enabling"? I imagine that in most cases the person wants to
> enable everything, though providing that does make it easier for people to
> turn on lots of flags which may be against our long-term philosophy of
> trying to reduce flag usage...
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/13808#issuecomment-257862556>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAazfGvLPYrqrZ1qEJ9OKVH0B2hIpEU5ks5q6I7qgaJpZM4Km2FI>
> .
>

sheelio commented:
I'm guessing we only want this for superusers (since feature flags don't make sense for normal users anyway?)

(Alternatively, we could have it always copy the feature flags if you're not a superuser and you copy an app from a project with feature flags enabled).

czue commented:
looks like this is already superuser only

orangejenny commented:
How about checkboxes to enable flags? I like the thought of making that easier but am wary of copying all flags because of the ones that control things like the new exports backend and new cloudcare.

czue commented:
checkboxes was actually my first thought, but I thought that might be asking you to do too much work. :)

amsagoff commented:
superuser only and checkboxes all the way!



On Wed, Nov 2, 2016 at 11:23 AM, Cory Zue <notifications@github.com> wrote:

> checkboxes was actually my first thought, but I thought that might be
> asking you to do too much work. :)
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/13808#issuecomment-257898342>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAazfFfltY0q4hfFE__P6j93-rG9-FTBks5q6KsNgaJpZM4Km2FI>
> .
>

esoergel commented on corehq/apps/app_manager/templates/app_manager/v1/app_view.html:
```diff
@@ -188,6 +188,48 @@
+                $form = $submit.closest("form"),
+                domain = $form.find("#id_domain").val(),
+                $modal = $("#copy-feature-flags");
+            {% if request|toggle_enabled:"SUPPORT" %}
```
Shouldn't this be enabled for all dimagi users?  You could use `user.is_superuser`, since those are the users who can access feature flags.

esoergel commented on corehq/apps/app_manager/templates/app_manager/v1/app_view.html:
```diff
@@ -188,6 +188,48 @@
+                            });
+                            $modal.one("hide.bs.modal", function() {
+                                $submit.enableButton();
+                            });
```
This is cool, I hadn't seen [`.one`](http://api.jquery.com/one/) before.

esoergel commented on corehq/apps/domain/views.py:
```diff
@@ -2525,6 +2528,21 @@ def page_context(self):
+                if t.enabled(request.domain) and not t.enabled(other_domain)]
+        diff.sort(cmp=lambda x, y: cmp(x[1], y[1]))
+        diff = ["<a href='{}'>{}</a>".format(reverse(ToggleEditView.urlname, args=[d[0]]), d[1]) for d in diff]
+    return json_response(diff)
```
Probably not worth changing at this point, but you may be able to do this client-side with just [`toggles.js`](https://github.com/dimagi/commcare-hq/blob/f466e62b0754ca09c4dc0e3cc784cb0c9a31cd0f/corehq/apps/hqwebapp/views.py#L1119-L1126), which the user should already have cached in their browser.

orangejenny replied:
Would be nice to do client-side, but toggles.js only knows about the current domain's flags, not the one you're copying to.

esoergel replied:
You can access it for any domain the user has access to:
```python
In [1]: reverse('toggles_js', args=['mydomain'])
Out[1]: u'/a/mydomain/toggles.js'
```
Although I guess it's imported as a named module, so I could see that being tricky to get two copies in accessible formats.

orangejenny replied:
Oh, I see what you mean. Makes sense, but think I'm going to leave it as-is.

esoergel replied:
sounds good

esoergel reviewed (commented):
I had a couple minor, optional pieces of feedback, but implementation-wise, this looks g2m as-is.  From the PR discussion it looks like there'll be some changes though, so I'll hold off on approval for those.

orangejenny committed:
superuser, not support toggle

orangejenny committed:
Add front end checkboxes

orangejenny committed:
Merge branch 'master' into feature-flag-diff

orangejenny committed:
back end for toggle enabling

orangejenny committed:
include feature previews

orangejenny committed:
change language: 'feature flag' => 'toggle'

orangejenny commented:
Added checkboxes to selectively enable flags while coying. Also added feature previews, but kept it superuser-only - didn't think the cost of the extra step was worth the benefit for non-superusers, since there are relatively few feature previews.

<img width="654" alt="screen shot 2016-11-03 at 5 06 27 pm" src="https://cloud.githubusercontent.com/assets/1486591/19985227/065dcaf0-a1e8-11e6-9029-b4ad724ee8a1.png">

czue commented:
(added the announce label, for next internal announcement, because this is sweet)
